core/unit_test/TestExecSpaceThreadSafety.hpp

```diff
@@ -0,0 +1,539 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#include <Kokkos_Core.hpp>
+#include <thread>
+
+namespace {
+
+#ifdef KOKKOS_ENABLE_OPENMP
+template <class Lambda1, class Lambda2>
+void run_threaded_test(const Lambda1 l1, const Lambda2 l2) {
+#pragma omp parallel num_threads(2)
+  {
+    if (omp_get_thread_num() == 0) l1();
+    if (omp_get_thread_num() == 1) l2();
+  }
+}
+// We cannot run the multithreaded test when threads or HPX is enabled because
+// we cannot launch a thread from inside another thread
+#elif !defined(KOKKOS_ENABLE_THREADS) && !defined(KOKKOS_ENABLE_HPX)
+template <class Lambda1, class Lambda2>
+void run_threaded_test(const Lambda1 l1, const Lambda2 l2) {
+  std::thread t1(l1);
+  std::thread t2(l2);
+  t1.join();
+  t2.join();
+}
+#else
+template <class Lambda1, class Lambda2>
+void run_threaded_test(const Lambda1 l1, const Lambda2 l2) {
+  l1();
+  l2();
+}
+#endif
+
+// The idea for all of these tests is to access a View from kernels submitted by
+// two different threads to the same execution space instance. If the kernels
+// are executed concurrently, we expect to detect some indices in the View that
+// have unexpected values. We map the View element to work items in reversefor
+// for one of the threaads to try triggering problems better.
+void run_exec_space_thread_safety_range() {
+  constexpr int N = 1000;
+  constexpr int M = 1000;
+
+  Kokkos::View<int *, TEST_EXECSPACE> view("view", N);
+  Kokkos::View<Kokkos::pair<int, int>, TEST_EXECSPACE> error_index(
+      "error_index");
+
+  run_threaded_test(
+      [=]() {
+        TEST_EXECSPACE exec;
+        for (int j = 0; j < M; ++j) {
+          Kokkos::parallel_for(
+              "bar", Kokkos::RangePolicy<TEST_EXECSPACE>(exec, 0, N),
+              KOKKOS_LAMBDA(int i) {
+                // check that values in the View don't differ by more than one
+                // which would indicate that the other kernel has written to it
+                // while this one is running.
+                if (i < N - 1 && (Kokkos::atomic_load(&view(i + 1)) -
+                                  Kokkos::atomic_load(&view(i))) > 1)
+                  Kokkos::atomic_store(error_index.data(),
+                                       Kokkos::pair{j + 1, i});
+                Kokkos::atomic_inc(&view(i));
```

Comment:
Why did you choose to increment value rather than set them to different distinguished values in the two kernels?
The incrementing sanity check logic is more complex.